packages/framework/ai-collab/src/explicit-strategy/promptGeneration.ts

```diff
@@ -0,0 +1,390 @@
+	Edits are JSON objects that conform to the following schema.
+	The top level object you produce is an "EditWrapper" object which contains one of "SetRoot", "Insert", "Modify", "Remove", "Move", or null.
+	${treeSchemaString}
+	The tree is a JSON object with the following schema: ${promptFriendlySchema}
+	${plan === undefined ? "" : `You have made a plan to accomplish the user's goal. The plan is: "${plan}". You will perform one or more edits that correspond to that plan to accomplish the goal.`}
+	${
+		log.length === 0
+			? ""
+			: `You have already performed the following edits:
+			${createEditList(log)}
+			This means that the current state of the tree reflects these changes.`
+	}
+	The current state of the tree is: ${decoratedTreeJson}.
+	${log.length > 0 ? "Before you made the above edits t" : "T"}he user requested you accomplish the following goal:
+	"${userPrompt}"
+	If the goal is now completed or is impossible, you should return null.
+	Otherwise, you should create an edit that makes progress towards the goal. It should have an English description ("explanation") of which edit to perform (specifying one of the allowed edit types).`;
+	return systemPrompt;
+}
+
+/**
+ * Generates a prompt designed to make an LLM review the edits it created and applied to a SharedTree based
+ * on a user-specified goal. This prompt is designed to give the LLM's ability to correct for mistakes and improve the accuracy/fidelity of its final set of tree edits
+ */
+export function getReviewSystemPrompt<TSchema extends ImplicitFieldSchema>(
+	userPrompt: string,
+	idGenerator: IdGenerator,
+	view: TreeView<TSchema>,
+	treeNode: TreeNode,
+	originalDecoratedJson: string,
+	appGuidance?: string,
+): string {
+	const isRootNode = Tree.parent(treeNode) === undefined;
+	const schema = isRootNode
+		? normalizeFieldSchema(view.schema)
+		: normalizeFieldSchema(Tree.schema(treeNode));
+	const promptFriendlySchema = getPromptFriendlyTreeSchema(getJsonSchema(schema.allowedTypes));
+	const decoratedTreeJson = toDecoratedJson(idGenerator, treeNode);
+
+	const role = `You are a collaborative agent who interacts with a JSON tree by performing edits to achieve a user-specified goal.${
+		appGuidance === undefined
+			? ""
+			: `
+			The application that owns the JSON tree has the following guidance: ${appGuidance}`
+	}`;
+
+	// TODO: security: user prompt in system prompt
+	const systemPrompt = `
+	${role}
+	You have performed a number of actions already to accomplish a user request.
+	You must review the resulting state to determine if the actions you performed successfully accomplished the user's goal.
+	The tree is a JSON object with the following schema: ${promptFriendlySchema}
+	The state of the tree BEFORE changes was: ${originalDecoratedJson}.
+	The state of the tree AFTER changes is: ${decoratedTreeJson}.
+	The user requested that the following goal should be accomplished:
+	${userPrompt}
+	Was the goal accomplished?`;
+	return systemPrompt;
+}
+
+/**
+ * TBD
```

Comment:
@noencke @taylorsw04 recommendations?

Reply:
(and for the others)

Reply:
This method translates from the JSON Schema representation to a textual, TypeScript-style set of types for embedding in a prompt.

Reply:
FWIW, I think we should have an item to kill this and use TypeChat or any of the numerous TS-generation libraries out there. No reason to build our own.